src/internal/pfsdb/common.go

```diff
@@ -0,0 +1,199 @@
+package pfsdb
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream"
+)
+
+type (
+	SortOrder      int
+	FilterOperator string
+)
+
+const (
+	SortNone SortOrder = iota
+	SortAscend
+	SortDescend
+)
+
+const (
+	// Note that we use `?` as the placeholder for bindvar, but postgres uses $1, $2, etc.
+	// We solve this by rebinding the query before executing it using sqlx.Rebind.
+	Equal              FilterOperator = "%s = ?"
+	GreaterThan        FilterOperator = "%s > ?"
+	GreaterThanOrEqual FilterOperator = "%s >= ?"
+	LessThan           FilterOperator = "%s < ?"
+	LessThanOrEqual    FilterOperator = "%s <= ?"
+	NotEqual           FilterOperator = "%s <> ?"
+	ValueIn            FilterOperator = "%s in (?)"
+)
+
+type (
+	ModelType interface{ Repo | Commit | Branch }
+	FieldType interface{ CommitField | BranchField }
+	PairType  interface{ CommitPair | BranchPair }
+)
+
+type Filter[T FieldType] struct {
```

Comment:
I have some concerns here.  One is that it is apparently invalid to use more than one field type in a query, which is not true in the presence of joins `(project.name = ? AND repo.name = ? AND branch.id = ?)`.  The other is mixing "code and data" here; the query builder needs to know what values are bound to the query in order to produce it?  That's pretty weird.

I am not convinced that `repo.name IN (...)` is an important use case to support (it's used in one test and the test isn't really running the right query), but it seems like `x = ANY $1` where $1 is an array does the same thing.